lib/Hakyll/Web/Template/Context.hs

```diff
@@ -306,6 +312,8 @@ getItemUTC locale id' = do
         [tryField "published" fmt | fmt <- formats] ++
         [tryField "date"      fmt | fmt <- formats] ++
         [parseTime' "%Y-%m-%d" $ intercalate "-" $ take 3 $ splitAll "-" fnCand | fnCand <- reverse paths]
+        [parseTime' "%Y-%m-%d" $ intercalate "-" $ fnCand 
+          | fnCand <- takeWhile ((==3) . length) . map (take 3) . iterate (drop 1) $ paths]
```

Comment:
Just make use of `Data.List`:
```suggestion
          | fnCand <- map (take 3) $ tails paths]
```
Also didn't you want a `reverse` in here, to prefer the rightmost candidate?